url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/LaneAgent.java

```diff
@@ -0,0 +1,320 @@
+package agent;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+
+import DeviceGraphics.DeviceGraphics;
+import DeviceGraphics.PartGraphics;
+import GraphicsInterfaces.LaneGraphics;
+import agent.data.Part;
+import agent.interfaces.Lane;
+import factory.PartType;
+
+/**
+ * Lane delivers parts to the nest
+ * @author Arjun Bhargava
+ */
+public class LaneAgent extends Agent implements Lane {
+
+	public List<PartType> requestList = Collections
+			.synchronizedList(new ArrayList<PartType>());
+	public List<MyPart> currentParts = Collections
+			.synchronizedList(new ArrayList<MyPart>());
+
+	public PartType currentType;
+	
+	public int topLimit = 5;
+	public int lowerThreshold = 3;
+	public int extraRequestCount = 0;
+
+	public LaneStatus state;
+
+	String name;
+
+	public Semaphore animation = new Semaphore(0, true);
+
+	public class MyPart {
+		public Part part;
+		PartStatus status;
+
+		public MyPart(Part p) {
+			part = p;
+			status = PartStatus.BEGINNING_LANE;
+		}
+	}
+
+	public enum PartStatus {
+		BEGINNING_LANE, IN_LANE, END_LANE, TOLD_NEST, NEED_TO_DELIVER, DELIVERED
+	};
+
+	public enum LaneStatus {
+		FILLING, PURGING, WAITING, BROKEN, BROKEN_WHILE_PURGING;
+	};
+
+	FeederAgent feeder;
+	NestAgent nest;
+	LaneGraphics laneGUI;
+
+	public LaneAgent(String name) {
+		super();
+
+		this.name = name;
+		state = LaneStatus.FILLING;
+	}
+
+	@Override
+	public void msgINeedPart(PartType type) {
+		print("Received msgINeedPart");
+		requestList.add(type);
+		currentType = type;
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgeParts() {
+		print("Received msgPurgeParts");
+		if(state == LaneStatus.BROKEN_WHILE_PURGING || state == LaneStatus.BROKEN) {
+			state = LaneStatus.BROKEN_WHILE_PURGING;
+		} else {
+			state = LaneStatus.PURGING;
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgHereIsPart(Part p) {
+		print("Received msgHereIsPart of type " + p.type.getName());
+		currentParts.add(new MyPart(p));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **89**. At this line the following field is not protected by synchronization: `currentParts`. Your code typically protects this field with synchronization on object: `currentParts`.

**Fix**  
 Consider protecting the above field with synchronization on object: `currentParts`. For example, like at lines: **191**, **116**, **288**, **290**, **182**, and **101**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **101**, **116**, **182**, **191**, **288**, and **290**.